### DISCUSSION TITLE
Context Propagation Inject Method Debugging

### DISCUSSION THREAD

**Discussion by paychex-ssmithrand** [open-telemetry/opentelemetry-ruby#1356] [Q&A] [answered]
Thank you to @ahayworth for answering my earlier question and getting me started on this path.

I've been developing a custom context propagator that will extract our custom trace headers and turn them into trace ids and span ids, use them internally, export them via OTLP to a collector, and then inject them as custom trace headers on the way out.

I've gotten to the point where my `extract ` method is successfully finding the headers and converting them into ids, but I cannot seem to ascertain whether my `inject` method is being called. I have some print statements in it that seem to never execute, which makes me wonder if I'm returning something invalid or missing a step in `extract`. I've looked at the various other propagators included in the core api as well as contrib, and I seem to be doing everything the same. Code below:

```
def extract(carrier, context: Context.current, getter: Common::Propagation.rack_env_getter)
          trace_id = getter.get(carrier, PAYX_TXID_HEADER)
          span_id  = getter.get(carrier, PAYX_REQID_HEADER)

          return context unless trace_id
          
          trace_id = to_trace_id(trace_id.gsub('-',''))
          span_id  = to_span_id(span_id.gsub('-','').truncate(16,omission:""))

          span_context = Trace::SpanContext.new(
            trace_id: trace_id,
            span_id: span_id,
            tracestate: nil,
            remote: true
          )
          p span_context.valid? #Prints true
          span = OpenTelemetry::Trace.non_recording_span(span_context)
          OpenTelemetry::Trace.context_with_span(span, parent_context: context)
        end

def inject(carrier, context: Context.current, setter: Context::Propagation.text_map_setter)
          span_context = OpenTelemetry::Trace.current_span(context).context
          p span.context.valid? #Does not print at all
          return unless span_context.valid?

          p span_context.hex_trace_id()
          p span_context.hex_span_id()
          p carrier.keys()

          setter.set(carrier, PAYX_TXID_HEADER, span_context.hex_trace_id())
          setter.set(carrier, PAYX_REQID_HEADER, span_context.hex_span_id())
          nil
        end
```

**Comment by plantfansam** [comment]
@paychex-ssmithrand have you tried stepping through the code with a debugger? Are you sure you're using instrumentation gems that call `inject` somewhere?

Some maybe helpful background on propagators: the OpenTelemetry spec requires, and this library implements, a _composite_ propagator. This runs each registered propagator whenever `extract` and `inject` are called. You can follow the code path starting from the [`Configurator`](https://github.com/open-telemetry/opentelemetry-ruby/blob/672739120a605c5c14afc48e5aa07d1320443355/sdk/lib/opentelemetry/sdk/configurator.rb#L215). It is the responsibility of _instrumentation_ to call [inject](https://github.com/open-telemetry/opentelemetry-specification/blob/494bc180dba9d89a19457454f25e57b0fc7bcd31/specification/context/api-propagators.md#inject) — in the context of this library, that's invoked with `OpenTelemetry.propagation.inject(carrier)`. See, for example, [the HTTP client library](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/22796cb8141b3100c6ce94c25a15e31df3313464/instrumentation/http_client/lib/opentelemetry/instrumentation/http_client/patches/client.rb#L40).

**Reply by paychex-ssmithrand** [reply]
~~I can confirm that the instrumentation I'm using (HTTP) is calling inject. If I remove my custom propagator and just use trace_context and baggage propagators, I end up seeing spans being exported to my collector. I am using my propagator like so:
EDIT: removing my propagator from the list actually **does not** produce spans, so I am wondering if I am passing in the correct things to `c.propagators`?~~

After looking things over, it appears that HTTP instrumentation is not catching my requests to this service at all, but Action_Pack is. I looked through Action_Pack instrumentation and I could not find a call to `inject`.

```
my_custom_propagator = OpenTelemetry::Propagator::Custom.text_map_propagator
trace_context_propagator = OpenTelemetry::Trace::Propagation::TraceContext.text_map_propagator
baggage_propagator =  OpenTelemetry::Baggage::Propagation.text_map_propagator

OpenTelemetry::SDK.configure do |c|
  c.service_name = 'my-service'
  c.propagators = [trace_context_propagator,baggage_propagator,my_custom_propagator]
  c.use 'OpenTelemetry::Instrumentation::HTTP'
end
```

**Reply by arielvalentin** [reply]
@paychex-ssmithrand Is it possible that your custom propagator is changing the sampling decision in some way? I see in the code snippet above you do not set the `trace_flags` attribute so it is setting it to `DEFAULT`[^1], which currently means sampling is turned `off`[^2]. 

[^1]: https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/trace/span_context.rb#L27
[^2]: https://github.com/open-telemetry/opentelemetry-ruby/blob/fcecf418eb67804c1bc6b1dfdf9ca71949f2f40f/api/lib/opentelemetry/trace/trace_flags.rb#L43

**Reply by paychex-ssmithrand** [reply]
@arielvalentin  Thank you, that got me a bit closer. I see spans being exported to my collector now (makes sense as nothing was being sampled so nothing would be sent to it). However, I am still not seeing the inject method being called - and I still think that is due to `Action_Pack` instrumentation not calling `inject`, but rather doing some internal propagation work?

**Reply by arielvalentin** [reply]
I'm afraid I don't understand what you are asking. 

Injecting headers should happen on outbound HTTP requests and responses. So you should only see inject being called in instrumentation for HTTP client libraries. 

Your application will _extract_ headers when it is the server. This happens in the Rack instrumentation. 

Are you having trouble with propagating context using the Inject or continuing traces using Extract?

**Reply by ahayworth** [reply]
This sounds like you're building something in Rails, and maybe you're expecting to see your custom headers returned when you inspect the web requests in the Chrome dev tools, or something. Like @arielvalentin said, that's not what context propagation injection is actually about.

Context propagation _`extract`_ is called when a server receives a request. It _extracts_ the necessary headers so that it has the information it needs from the client to continue a trace.

Context propagation _`inject`_ is called whenever a client _sends_ a request. It _injects_ the necessary headers so that the receiving server has the information it needs to continue a trace.

To test this: you could just write a script using the [Net::HTTP instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/net_http/example/net_http.rb) combined with your custom propagation, and all it needs to do is make a request to your Rails app. When you visualize the trace, you should be able to see the client (`Net::HTTP`) span as the logical parent of the server spans. If that relationship happens and the trace "looks right", you can presume that it's fully functional at that point. ❤️ 

---

Notably - none of our instrumentation calls `inject` when responding to a browser request (at least, none that I know of right now). Typically there's no reason for us to actually do that, because even if the browser happens to be doing OpenTelemetry tracing, it doesn't need or want to continue a trace using the server's span info; that would make further work on the client appear (in a trace) as though it was a logical child of the server's operation which isn't actually the case. 😄

**Reply by paychex-ssmithrand** [reply]
Sorry for the radio silence @arielvalentin and @ahayworth. Our in-house propagation libraries were built with the concept of adding response trace headers even though it is redundant as the information has been recorded as the request comes in. I understand now what is going on.

Things seem to be working correctly!

I am now on to converting baggage to span attributes so that the collector can see the additional contextual data that comes along via custom headers. Of course, if this is conceptually not correct, please do let me know.

**Comment by paychex-ssmithrand** [comment, marked as answer]
Quoting @arielvalentin and @ahayworth from the above thread as the answer:

> Injecting headers should happen on outbound HTTP requests and responses. So you should only see inject being called in instrumentation for HTTP client libraries.
> Your application will extract headers when it is the server. This happens in the Rack instrumentation.

> This sounds like you're building something in Rails, and maybe you're expecting to see your custom headers returned when you inspect the web requests in the Chrome dev tools, or something. Like @arielvalentin said, that's not what context propagation injection is actually about.

> Context propagation extract is called when a server receives a request. It extracts the necessary headers so that it has the information it needs from the client to continue a trace.

> Context propagation inject is called whenever a client sends a request. It injects the necessary headers so that the receiving server has the information it needs to continue a trace.

> To test this: you could just write a script using the [Net::HTTP instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/net_http/example/net_http.rb) combined with your custom propagation, and all it needs to do is make a request to your Rails app. When you visualize the trace, you should be able to see the client (Net::HTTP) span as the logical parent of the server spans. If that relationship happens and the trace "looks right", you can presume that it's fully functional at that point. ❤️

> Notably - none of our instrumentation calls inject when responding to a browser request (at least, none that I know of right now). Typically there's no reason for us to actually do that, because even if the browser happens to be doing OpenTelemetry tracing, it doesn't need or want to continue a trace using the server's span info; that would make further work on the client appear (in a trace) as though it was a logical child of the server's operation which isn't actually the case. 😄
